File: bilang/src/java/bilang/SymbolTable.java
package bilang;

import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

class SymbolTable<T> {
    final private LinkedList<Map<String, T>> scope;

    SymbolTable(Map<String, T> initial) {
        this.scope = new LinkedList<>(Collections.singletonList(initial));
    }

    void push() {
        scope.addFirst(new HashMap<>());
    }

    void pop() {
        scope.removeFirst();
    }

    Map<String, T> currentScope() {
        return scope.peekFirst();
    }

    T lookup(String v) {
        for (Map<String, T> m : scope) {
            if (m.containsKey(v))
                return m.get(v);
        }
        return null;
    }

    void update(String v, T val) {
        for (Map<String, T> m : scope) {
            if (m.containsKey(v)) {
                m.put(v, val);
                return;
            }
        }
        throw new RuntimeException("Undefined variable " + v);
    }

    @Override
    public String toString() {
        return scope.toString();
    }
}

File: bilang/src/bilang/SymbolTable.kt
package bilang

import java.util.*

internal class SymbolTable<T>(initial: MutableMap<String, T>) {
    private val scope = LinkedList<MutableMap<String, T>>(listOf(initial))

    fun push() {
        scope.addFirst(mutableMapOf())
    }

    fun pop() {
        scope.removeFirst()
    }

    fun currentScope(): MutableMap<String, T> = scope.peekFirst()

    fun lookup(v: String): T? {
        for (m in scope) {
            if (m.containsKey(v))
                return m[v]
        }
        return null
    }

    fun update(v: String, value: T) {
        for (m in scope) {
            if (m.containsKey(v)) {
                m[v] = value
                return
            }
        }
        throw RuntimeException("Undefined variable $v")
    }

    override fun toString() = scope.toString()
}